src/parallel/communicator.C

```diff
@@ -197,27 +213,40 @@ void Communicator::barrier () const
 void Communicator::barrier () const {}
 #endif
 
-
-MessageTag Communicator::get_unique_tag(int tagvalue) const
+#ifdef LIBMESH_HAVE_MPI
+void Communicator::nonblocking_barrier (Request & req) const
 {
-  if (used_tag_values.count(tagvalue))
+  if (this->size() > 1)
     {
-      // Get the largest value in the used values, and pick one
-      // larger
-      tagvalue = used_tag_values.rbegin()->first+1;
-      libmesh_assert(!used_tag_values.count(tagvalue));
+      LOG_SCOPE("nonblocking_barrier()", "Parallel");
+      libmesh_call_mpi(MPI_Ibarrier (this->get(), req.get()));
     }
-  used_tag_values[tagvalue] = 1;
+}
+#else
+void Communicator::nonblocking_barrier (Request & /*req*/) const {}
+#endif
+
+
+MessageTag Communicator::get_unique_tag(int /*tagvalue*/) const
+{
+  // Don't give out tag values greater than what MPI can handle!
+  if (_next_tag == _max_tag)
+    _next_tag = 0;
+
+  auto new_tag = _next_tag++;
+
+  // We don't want to simply get the largest one
+  // because it might be near MPI_TAG_UB and then
+  // we can end up in a sticky situation
+  // Most-likely we will amost never hit a
+  // tag that is in use because we have
+  // ~2 billion of them
+  while(used_tag_values.count(new_tag))
+    new_tag++;
```

Comment:
Ignoring tagvalue prevents use cases where the user wants to request tags in different orders on different processors; e.g. to use a task management system where ranks grab new bits of work as they complete prior tasks.  We don't actually *do* that or intend to at any time in the near future, but it was the original reasoning.

So what's the reasoning behind removing it?  We already had a "grab the next larger tag if the one we looked at first is already in use" mechanism.  Why didn't it work for your case?